sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -261,6 +262,78 @@ protected Symbol visitFunctionCall(FunctionCall node, ExpressionAnalysisContext
         }
 
         @Override
+        protected Symbol visitSimpleCaseExpression(SimpleCaseExpression node, ExpressionAnalysisContext context) {
+            return createCaseFunction(node.getWhenClauses(), node.getOperand(), node.getDefaultValue(), context);
+        }
+
+        @Override
+        protected Symbol visitSearchedCaseExpression(SearchedCaseExpression node, ExpressionAnalysisContext context) {
+            return createCaseFunction(node.getWhenClauses(), null, node.getDefaultValue(), context);
+        }
+
+        private Symbol createCaseFunction(List<WhenClause> whenClauseExpressions,
+                                          @Nullable Expression operandExpression,
+                                          @Nullable Expression defaultValue,
+                                          ExpressionAnalysisContext context) {
+            List<Symbol> operands = new ArrayList<>(whenClauseExpressions.size());
+            List<Symbol> results = new ArrayList<>(whenClauseExpressions.size());
+            Set<DataType> resultsTypes = new HashSet<>(whenClauseExpressions.size());
+
+            // check for global operand
+            Symbol operandLeftSymbol = null;
+            if (operandExpression != null) {
+                operandLeftSymbol = operandExpression.accept(innerAnalyzer, context);
+            }
+
+            for (WhenClause whenClause : whenClauseExpressions) {
+                Symbol operandRightSymbol = whenClause.getOperand().accept(innerAnalyzer, context);
+                Symbol operandSymbol = operandRightSymbol;
+
+                if (operandLeftSymbol != null) {
+                    FunctionIdent ident = new FunctionIdent(EqOperator.NAME,
+                        Arrays.asList(operandLeftSymbol.valueType(), operandRightSymbol.valueType()));
+                    operandSymbol = context.allocateFunction(
+                        new FunctionInfo(ident, EqOperator.RETURN_TYPE),
+                        Arrays.asList(operandLeftSymbol, operandRightSymbol)
+                    );
+                }
+
+                operands.add(operandSymbol);
+
+                Symbol resultSymbol = whenClause.getResult().accept(innerAnalyzer, context);
+                results.add(resultSymbol);
+                resultsTypes.add(resultSymbol.valueType());
+            }
+
+            if (resultsTypes.size() == 2 && !resultsTypes.contains(DataTypes.UNDEFINED)
+                || resultsTypes.size() > 2) {
+                throw new UnsupportedOperationException(
+                    "Data types of all result expressions of a CASE statement must be equal");
```

Comment:
maybe output the types that have been found as well for easier debugging for the user